### PR TITLE
Filestore improvements

### PR DESCRIPTION
**Pull request by kdmccormick**
@edx/masters-neem 

Another pre-cursor for both https://openedx.atlassian.net/browse/EDUCATOR-4335 and https://openedx.atlassian.net/browse/EDUCATOR-4377.

### CODE REVIEW

**Review by zacharis278 — APPROVED** [review]
👍

File: registrar/apps/core/filestore.py [hunk 1]
```diff
@@ -7,31 +11,91 @@
...
         self.backend = backend
+        self.path_prefix = path_prefix
 
     def store(self, path, contents):
         """
         Store a file.
 
         Arguments:
             path: Path to file within filestore.
+                Will be prefixed by `path_prefix`.
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: Nit: prefixed by `self.path_prefix`

File: registrar/apps/core/filestore.py [hunk 2]
```diff
@@ -7,31 +11,91 @@
...
             URL to result
         """
-        self.backend.save(path, ContentFile(bytes(contents, 'utf-8')))
+        full_path = self.get_full_path(path)
+        self.backend.save(full_path, ContentFile(bytes(contents, 'utf-8')))
         return self.get_url(path)
 
+    def retrieve(self, path):
+        """
+        Store a file.
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: Nit: "Retrieve a file"

Comment (iloveagent57) [reply registrar/apps/core/filestore.py]: Or maybe "retrieve the contents of a file"

File: registrar/apps/core/filestore.py [hunk 3]
```diff
@@ -7,31 +11,91 @@
...
+                Will be prefixed by `path_prefix`.
+
+        Returns: bool
+        """
+        full_path = self.get_full_path(path)
+        return self.backend.exists(full_path)
+
+    def get_full_path(self, path):
+        """
+        Apply `path_prefix` to `path`. Use POSIX path joining.
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: Nit: `self.path_prefix`

File: registrar/apps/core/filestore.py [hunk 4]
```diff
@@ -7,31 +11,91 @@
...
+        """
+        return posixpath.join(self.path_prefix, path)
+
     def get_url(self, path):
         """
         Given the path of a file in the store, return a URL to the file.
+        Path will be prefixed by `path_prefix`.
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: Nit: `self.path_prefix`

Comment (iloveagent57) [reply registrar/apps/core/filestore.py]: Actually, you're talking about implementation here for a method that has no implementation - maybe just nix this line?

File: registrar/apps/core/filestore.py [hunk 5]
```diff
@@ -46,8 +110,9 @@ class FileSystemFilestore(FilestoreBase):
...
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: `self.path_prefix`

File: registrar/apps/core/filestore.py [hunk 6]
```diff
@@ -63,27 +128,28 @@ def __init__(self, *args, **kwargs):
...
```
Comment (iloveagent57) [inline registrar/apps/core/filestore.py]: `self.path_prefix`

File: registrar/apps/core/jobs.py [hunk 7]
```diff
@@ -20,16 +20,17 @@
...
 
+from registrar.apps.core.constants import JOB_RESULT_PATH_PREFIX
 from registrar.apps.core.filestore import get_filestore
 from registrar.apps.core.permissions import JOB_GLOBAL_READ
 
 
 JobStatus = namedtuple('JobStatus', ['created', 'state', 'result'])
 
 logger = logging.getLogger(__name__)
+filestore = get_filestore(JOB_RESULT_PATH_PREFIX)
```
Comment (iloveagent57) [inline registrar/apps/core/jobs.py]: maybe `result_filestore`?

File: registrar/apps/core/tests/test_filestore.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        cls._s3_mock.stop()
+        super().tearDownClass()
+
+    prefixes = ("", "prefix", "prefix/withslashes/")
+    paths = ("file.txt", "folder/file.txt")
+    contents = ("filecontents!", "")
+
+    @ddt.data(*(product(prefixes, paths, contents)))
+    @ddt.unpack
+    def test_s3_filestore(self, prefix, path, contents):
```
Comment (iloveagent57) [inline registrar/apps/core/tests/test_filestore.py:47]: Nit: you're reusing the `contents` variable name that's fined in a different scope.

**Review by iloveagent57 — APPROVED** [review]
Looks good, just a bunch of nits.

File: registrar/apps/core/filestore.py (hunk 4 above)
Comment (kdmccormick) [reply registrar/apps/core/filestore.py]: My thinking is that all implementations will (or should) follow that logic, so it's worth noting in the base class. I changed the wording from "Path will be" to "Path should be".

**Pull request merged**